File: src/lfsr.js
/**
 * @desc GB/NES style 15-bit LFSR (Linear Feedback Shift Register)
 */
class Lfsr {
  constructor () {
    this.__mode = 0
    this.__tapB = 1
    this.__register = 1
  }

  /**
   * @desc Set mode flag of LFSR tap. True: Use 6 / False: Use 1
   * @param {Boolean} mode 
   */
  setMode (mode) {
    this.__mode = mode
    if (mode) this.__tapB = 6
    else this.__tapB = 1
  }

  /**
   * @desc Calculate next bit
   */
  clock () {
    const bitA = this.__register & 1 // 1 is tabA
    const bitB = (this.__register >> this.__tapB) & 1
    const feedback = (bitA ^ bitB) << 14
    this.__register >>= 1
    this.__register |= feedback
  }

  /**
   * @desc Get hexadecimal random number from the latest 4-bits of register
   * @return {Number} Hexadecimal number (0 to 15)
   */
  getHex () {
    return this.__register >> 11
  }
}

export default Lfsr

File: src/subchips/lfsr.rs
pub struct LFSR {
    tap_b: u8,
    lfsr: u16
}

impl LFSR {
    pub fn new() -> LFSR {
        LFSR {
            tap_b: 1,
            lfsr: 1,
        }
    }

    pub fn reset(&mut self) {
        self.tap_b = 1;
        self.lfsr = 1;
    }

    pub fn clock(&mut self) {
        let bit_a = self.lfsr & 1;
        let bit_b = (self.lfsr >> self.tap_b) & 1;
        let feedback = (bit_a ^ bit_b) << 14;
        self.lfsr >>=1;
        self.lfsr |= feedback;
    }

    pub fn read_hex(&self) -> u8 {
        (self.lfsr & 0xF) as u8
    }

    pub fn set_noise_type(&mut self, noise_type: u32) {
        self.tap_b = match noise_type % 2 {
            0 => 1,
            1 => 6,
            _ => unreachable!()
        };
    }
}
